parquet/src/test/java/org/apache/iceberg/parquet/TestBloomRowGroupFilter.java

```diff
@@ -0,0 +1,950 @@
+    Assert.assertTrue("Should read: bloom filter doesn't help", shouldRead);
+
+    shouldRead = new ParquetBloomRowGroupFilter(SCHEMA, startsWith("required", "reqs"))
+        .shouldRead(parquetSchema, rowGroupMetadata, bloomStore);
+    Assert.assertTrue("Should read: bloom filter doesn't help", shouldRead);
+
+    shouldRead = new ParquetBloomRowGroupFilter(SCHEMA, startsWith("some_nulls", "somex"))
+        .shouldRead(parquetSchema, rowGroupMetadata, bloomStore);
+    Assert.assertTrue("Should read: bloom filter doesn't help", shouldRead);
+
+    shouldRead = new ParquetBloomRowGroupFilter(SCHEMA, startsWith("no_nulls", "xxx"))
+        .shouldRead(parquetSchema, rowGroupMetadata, bloomStore);
+    Assert.assertTrue("Should read: bloom filter doesn't help", shouldRead);
+  }
+
+  @Test
+  public void testMissingColumn() {
+    TestHelpers.assertThrows("Should complain about missing column in expression",
+        ValidationException.class, "Cannot find field 'missing'",
+        () -> new ParquetBloomRowGroupFilter(SCHEMA, lessThan("missing", 5))
+            .shouldRead(parquetSchema, rowGroupMetadata, bloomStore));
+  }
+
+  @Test
+  public void testColumnNotInFile() {
+    Expression[] exprs = new Expression[]{
+        lessThan("not_in_file", 1.0f), lessThanOrEqual("not_in_file", 1.0f),
+        equal("not_in_file", 1.0f), greaterThan("not_in_file", 1.0f),
+        greaterThanOrEqual("not_in_file", 1.0f), notNull("not_in_file"),
+        isNull("not_in_file"), notEqual("not_in_file", 1.0f), in("not_in_file", 1.0f, 2.0f)
+    };
+
+    for (Expression expr : exprs) {
+      boolean shouldRead = new ParquetBloomRowGroupFilter(SCHEMA, expr)
+          .shouldRead(parquetSchema, rowGroupMetadata, bloomStore);
+      Assert.assertTrue("Should read: bloom filter cannot be found: " + expr, shouldRead);
+    }
+  }
+
+  @Test
+  public void testColumnNotBloomFilterEnabled() {
+    Expression[] exprs = new Expression[]{
+        lessThan("non_bloom", "a"), lessThanOrEqual("non_bloom", "a"), equal("non_bloom", "a"),
+        greaterThan("non_bloom", "a"), greaterThanOrEqual("non_bloom", "a"), notNull("non_bloom"),
+        isNull("non_bloom"), notEqual("non_bloom", "a")
+    };
+
+    for (Expression expr : exprs) {
+      boolean shouldRead = new ParquetBloomRowGroupFilter(SCHEMA, expr)
+          .shouldRead(parquetSchema, rowGroupMetadata, bloomStore);
+      Assert.assertTrue("Should read: bloom filter cannot be found: " + expr, shouldRead);
+    }
+  }
+
+  @Test
+  public void testMissingStats() {
+    boolean shouldRead = new ParquetBloomRowGroupFilter(SCHEMA, equal("no_stats", "a"))
+        .shouldRead(parquetSchema, rowGroupMetadata, bloomStore);
+    Assert.assertFalse("Should skip: stats are missing but bloom filter is present", shouldRead);
+  }
+
+  @Test
+  public void testNot() {
+    // this test case must use a real predicate, not alwaysTrue(), otherwise binding will simplify it out
```

Comment:
would you mind explaining it?

Reply:
`TestBloomRowGroupFilter` follows the test style of `TestMetricsRowGroupFilter` and `TestDictionaryRowGroupFilter` and this comment is actually copied. I think what it means is that a real predicate e.g. equal, lessThanOrEqual, etc. need to be used. If you use `alwaysTrue`, it will be simplified out and then the visitEvaluator path won't be tested.